controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -339,12 +339,46 @@ func (r *BareMetalHostReconciler) credentialsErrorResult(err error, request ctrl
 }
 
 // hasRebootAnnotation checks for existence of reboot annotations and returns true if at least one exist
-func hasRebootAnnotation(host *metal3v1alpha1.BareMetalHost) bool {
-	for annotation := range host.Annotations {
+func hasRebootAnnotation(info *reconcileInfo) (hasReboot bool, rebootMode metal3v1alpha1.RebootMode) {
+	// Assume there's no RebootAnnotation until we check
+	rebootMode = metal3v1alpha1.RebootModeNone
+	for annotation, value := range info.host.GetAnnotations() {
+		// Don't use a break here as we may have multiple clients setting
+		// reboot annotations and we always want hard requests honoured
 		if isRebootAnnotation(annotation) {
-			return true
+			hasReboot = true
+			if isHardReboot(value, info) {
+				rebootMode = metal3v1alpha1.RebootModeHard
+			} else {
+				rebootMode = metal3v1alpha1.RebootModeSoft
+			}
 		}
 	}
+	return
+}
+
+// isHardReboot checks if reboot annotation requests a hard reboot and returns true if so
+func isHardReboot(annotation string, info *reconcileInfo) bool {
+
+	type RebootAnnotationArguments struct {
+		Mode	metal3v1alpha1.RebootMode	`json:"mode"`
+	}
```

Comment:
why this type being declared twice?
can't we just use `RebootAnnotationArguments` that was declared in `baremetalhost_types.go`?

Reply:
Doh! You're right, this is a carry over from my testing. Fixed.